core/arch/arm/kernel/thread.c

```diff
@@ -91,27 +93,44 @@ struct thread_core_local thread_core_local[CFG_TEE_CORE_NB_CORE] __nex_bss;
 #define STACK_CANARY_SIZE	0
 #endif
 
+#ifdef CFG_CORE_DEBUG_CHECK_STACKS
+/*
+ * Extra space added to each stack in order to reliably detect and dump stack
+ * overflows. Should cover the maximum expected overflow size caused by any C
+ * function (say, 512 bytes; no function should have that much local variables),
+ * plus the maximum stack space needed by __cyg_profile_func_exit(): about 1 KB,
+ * a large part of which is used to print the call stack. Total: 1.5 KB.
+ */
+#define STACK_CHECK_EXTRA	1536
+#else
+#define STACK_CHECK_EXTRA	0
+#endif
+
 #define DECLARE_STACK(name, num_stacks, stack_size, linkage) \
 linkage uint32_t name[num_stacks] \
-		[ROUNDUP(stack_size + STACK_CANARY_SIZE, STACK_ALIGNMENT) / \
-		sizeof(uint32_t)] \
+		[ROUNDUP(stack_size + STACK_CANARY_SIZE + STACK_CHECK_EXTRA, \
+			 STACK_ALIGNMENT) / sizeof(uint32_t)] \
 		__attribute__((section(".nozi_stack." # name), \
 			       aligned(STACK_ALIGNMENT)))
 
 #define STACK_SIZE(stack) (sizeof(stack) - STACK_CANARY_SIZE / 2)
 
-#define GET_STACK(stack) \
-	((vaddr_t)(stack) + STACK_SIZE(stack))
+#define GET_STACK(stack) ((vaddr_t)(stack) + STACK_SIZE(stack))
 
 DECLARE_STACK(stack_tmp, CFG_TEE_CORE_NB_CORE, STACK_TMP_SIZE, static);
 DECLARE_STACK(stack_abt, CFG_TEE_CORE_NB_CORE, STACK_ABT_SIZE, static);
 #ifndef CFG_WITH_PAGER
 DECLARE_STACK(stack_thread, CFG_NUM_THREADS, STACK_THREAD_SIZE, static);
 #endif
 
+#define GET_STACK_TOP_HARD(stack, n) \
+	((vaddr_t)&(stack)[n] + STACK_CANARY_SIZE / 2)
+#define GET_STACK_TOP_SOFT(stack, n) \
+	(GET_STACK_TOP_HARD(stack, n) + STACK_CHECK_EXTRA)
+#define GET_STACK_BOTTOM(stack, n) ((vaddr_t)&(stack)[n] + STACK_SIZE(stack[n]))
```

Comment:
No need to account for the canary at the bottom?

Reply:
I don't think so, it is taken into account in `STACK_SIZE()`.
```
[Lower addresses on the left]

| STACK_CANARY_SIZE/2 | STACK_CHECK_EXTRA | STACK_XXX_SIZE | STACK_CANARY_SIZE/2 |
^                     ^                   ^                ^
|                     |                   |                |
stack_xxx[n]          "hard" top          "soft" top       bottom
```

`STACK_SIZE` might not be a very good name. I might as well expand it in `GET_STACK_BOTTOM()`:
```C
#define GET_STACK_BOTTOM(stack, n) ((vaddr_t)&(stack)[n] + sizeof(stack[n]) - STACK_CANARY_SIZE / 2)
```

Do we need to force alignment to `STACK_ALIGNMENT` also? Or is it naturally the case? I think it is since alignment is taken into account in in `DECLARE_STACK()` and `STACK_CANARY_SIZE/2` is also a multiple of `STACK_ALIGNMENT`.

Reply:
`STACK_CANARY_SIZE` is chosen to be twice the required alignment of the stack so it should work automatically.

OK, I see that `GET_STACK_BOTTOM()` is correct now. I agree `STACK_SIZE()` isn't very good, do as you think is best.

By the way, that picture is really nice. Would you mind including it in the code?

Reply:
Sure.